File: python/Kth-Largest-Element-In-An-Array/Kth-Largest-Element-In-An-Array.py
## Python Solution 1:
class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        res = 0
        n = len(nums)
        if n==0 or k>n:
            return res
        nums = sorted(nums)
        res = nums[n-k]
        return res

      
 ### Python Solution 2:
class Solution:
    def findKthLargest(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        return heapq.nlargest(k, nums)[-1]
      
      
      
      
 ### Python Solution 3:
class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        res = 0
        n = len(nums)
        if n==0 or k>n:
            return 0
        start = 0
        end = n-1
        index = self.partition(nums,start,end)
        while index != n-k:
            if index > n-k:
                end = index - 1
                index = self.partition(nums,start,end)
            else:
                start = index + 1
                index = self.partition(nums,start,end)
        res = nums[n-k]
        return res

    def partition (self,nums: List[int],start:int,end:int) -> int:
        index = random.randint(start, end)
        pivot = nums[index]
        nums[index],nums[start] = nums[start],nums[index]    ##基准值放在数组开始
        pos = start
        for i in range (start+1,end+1):
            if nums[i]<pivot:
                pos+=1
                if i!=pos:
                    nums[pos],nums[i] = nums[i],nums[pos]
        nums[pos],nums[start] = nums[start],nums[pos]
        return pos
    
    # def partition (self,nums: List[int],start:int,end:int) -> int:
    #     index = random.randint(start, end)
    #     pivot = nums[index]
    #     nums[start],nums[index] = nums[index],nums[start]
    #     while start<end:
    #         while start<end and nums[end]>=pivot:
    #             end-=1
    #         nums[start] = nums[end]
    #         while start<end and nums[start]<pivot:
    #             start+=1
    #         nums[end] = nums[start]
    #     nums[start] = pivot
    #     return start

File: cpp/Kth-Largest-Element-In-An-Array/Kth-Largest-Element-In-An-Array.cpp
// C++ Solution 1:
class Solution {
public:
    int findKthLargest(vector<int>& nums, int k) {
        int res = 0;
        int n = nums.size();
        if (nums.empty() || k>n) return res;
        sort(nums.begin(),nums.end());
        res = nums[n-k];
        return res;        
    }
};

// C++ Solution 2:
class Solution {
public:
    int findKthLargest(vector<int>& nums, int k) {
        int res = 0;
        int n = nums.size();
        if (nums.empty() || k>n) return res;
        priority_queue<int, vector<int>,greater<int>> q;
        for (auto num:nums)
        {
            q.push(num);
            if (q.size()>k)
            {
                q.pop();
            }
        }
        res = q.top();
        return res;        
    }
};

// C++ Solution 3:
class Solution {
public:
    int findKthLargest(vector<int>& nums, int k) {
        int res = 0;
        int n = nums.size();
        if (nums.empty() || k>n) return res;
        int start = 0;
        int end = n -1;
        int index = Partition(nums,start,end);
        while(index!=n-k)
        {
            if (index>n-k)
            {
                end = index - 1;
                index = Partition(nums,start,end);
            }
            else
            {
                start = index+1;
                index = Partition(nums,start,end);
            }
        }
        res = nums[n-k];
        return res;        
    }
    int Partition(vector<int> &nums,int start,int end){
        int index = start + rand()%(end-start+1);
        int pivot = nums[index];
        swap(nums[index],nums[start]);
        while (start < end)
        {
            while(start < end && nums[end]>=pivot)
            {
                end--;
            }
            nums[start] = nums[end];
            while(start < end && nums[start]<pivot)
            {
                start++;
            }
            nums[end] = nums[start];
        }
        nums[start] = pivot;   // 基准值归位
        return start;
    }
    
    int Partition (vector<int> &nums,int start,int end){
        int index = start + rand()%(end-start+1) ;
        int pivot = nums[index];
        swap(nums[index],nums[start];
        int pos = start;
        for (int i=start+1;i<=end;i++)
        {
            if (nums[i]<pivot)
            {
                pos++;
                if (i!=pos)
                {
                    swap(nums[pos],nums[i]);
                }
            }
        }
        swap(nums[pos],nums[start]);
        return pos;
    }
};
